File: app/src/main/java/com.example.safehome/Door/DType.java
package com.example.safehome.Door;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * DType policies defines door types
 */
public enum DType {
    BIFOLD("BIFOLD"),
    DOUBLE("DOUBLE"),
    SINGLE("SINGLE"),
    POCKET("POCKET"),
    OPEN("OPEN"),
    SLIDING("SLIDING");

    private static final Logger log = LoggerFactory.getLogger(DType.class);
    private final String label;
    // static inference
    private static final Map<String, DType> BY_LABEL = new HashMap<>();
    /**
     * Create new SType with label = label
     * @param label
     */
    DType(String label){
        this.label = label;
    }
    /**
     * static inference collection of sensor types
     */
    static {
        for (DType e: values()) {
            BY_LABEL.put(e.label, e);
        }
        log.info(BY_LABEL.values().stream().collect(Collectors.toList()).toString());
    }
    /**
     * Static inference: Creates a SType Object with label = label.
     * @param label
     * @return
     */
    public static DType valueOfLabel(String label) {
        DType t = null;
        for (DType type : values()) {
            if (type.label.equals(label)) {
                log.info(type.label);
                t = type;
                return type;
            }
        }
        return t;
    }
//    public static void main(String[] args) {
//        DType type = DType.valueOfLabel("DOOR");
//        System.out.println("TYPE_" + type);
//    }
}

File: app/src/main/kotlin/com.example.safehome/Door/DType.kt
package com.example.safehome.Door

import com.example.safehome.Door.DType
import com.example.safehome.Door.Door
import com.example.safehome.Door.OpenDoor
import com.example.safehome.Door.BiFoldDoor
import com.example.safehome.Door.DoubleDoor
import com.example.safehome.Door.PocketDoor
import com.example.safehome.Door.SingleDoor
import com.example.safehome.AbstractFactory
import com.example.safehome.Door.SlidingDoor
import com.example.safehome.Door.DoorFactory
import com.example.safehome.Door.OpenDoorFactory
import com.example.safehome.Door.BiFoldDoorFactory
import com.example.safehome.Door.DoubleDoorFactory
import com.example.safehome.Door.PocketDoorFactory
import com.example.safehome.Door.SingleDoorFactory
import com.example.safehome.Door.SlidingDoorFactory
import com.example.safehome.Wall.Coordinates
import com.example.safehome.Wall.WallSegment
import com.example.safehome.Wall.Wall
import com.example.safehome.Camera.CameraMomento
import com.example.safehome.Momento
import com.example.safehome.Camera.SingletonCamera
import com.example.safehome.Camera.AbstractCamera
import android.provider.ContactsContract.CommonDataKinds.Phone
import com.example.safehome.System.SingletonSystem
import com.example.safehome.Sensors.SType
import com.example.safehome.Sensors.DoorSensor
import com.example.safehome.Sensors.MotionSensor
import com.example.safehome.Sensors.WindowSensor
import com.example.safehome.Sensors.SensorFactory
import com.example.safehome.Sensors.DoorSensorFactory
import com.example.safehome.Sensors.MotionSensorFactory
import com.example.safehome.Sensors.WindowSensorFactory
import com.example.safehome.FloorPlan.FType
import com.example.safehome.FloorPlan.FPlan
import com.example.safehome.FloorPlan.FloorPlan
import com.example.safehome.Publisher.CameraNotification
import com.example.safehome.Publisher.NotificationClient
import com.example.safehome.Publisher.SensorNotification
import com.example.safehome.ControlPanel.KeyPad
import com.example.safehome.ControlPanel.ControlPanel
import com.example.safehome.ControlPanel.ControlPanelMomento
import org.slf4j.LoggerFactory
import java.util.HashMap
import java.util.stream.Collectors

/**
 * DType policies defines door types
 */
enum class DType
/**
 * Create new SType with label = label
 * @param label
 */(  // static inference
    private val label: String
) {
    BIFOLD("BIFOLD"), DOUBLE("DOUBLE"), SINGLE("SINGLE"), POCKET("POCKET"), OPEN("OPEN"), SLIDING("SLIDING");

    companion object {
        private val log = LoggerFactory.getLogger(DType::class.java)
        private val BY_LABEL: MutableMap<String, DType> = HashMap()

        /**
         * static inference collection of sensor types
         */
        init {
            for (e in values()) {
                BY_LABEL[com.example.safehome.Door.e.label] = com.example.safehome.Door.e
            }
            log.info(BY_LABEL.values.stream().collect(Collectors.toList()).toString())
        }

        /**
         * Static inference: Creates a SType Object with label = label.
         * @param label
         * @return type
         */
        fun valueOfLabel(label: String): DType? {
            var t: DType? = null
            for (type in values()) {
                if (type.label == label) {
                    log.info(type.label)
                    t = type
                    return type
                }
            }
            return t
        } //    public static void main(String[] args) {
        //        DType type = DType.valueOfLabel("DOOR");
        //        System.out.println("TYPE_" + type);
        //    }
    }
}